lib/provisioner.js

```diff
@@ -0,0 +1,228 @@
+'use strict';
+
+require('./shims');
+
+const Iterate = require('taskcluster-lib-iterate');
+const {WMObject, errors} = require('./base');
+const {Bid} = require('./bid');
+
+const {buildWorkerConfiguration} = require('./worker-config');
+
+// We give the providers a proxied version of the worker configuration which
+// ensures that when the provider evaluates a worker configuration, only the
+// relevant fields are provided.  We do this because each provider needs to
+// run the worker configuration itself.  We don't want to allow providers
+// to use the incorrect section of the worker type configuration.
+let providerDataProxyHandler = {
+  get: function (target, prop, receiver) {
+    if (prop === 'evaluate') {
+      return function(...args) {
+        let {workerType, providerData} = target.evaluate(...args);
+        return {workerType, providerData};
+      }.bind(target);
+    } else {
+      return Reflect.get(...arguments);
+    }
+  },
+};
+
+let bidProxyHandler = {
+  get: function (target, prop, receiver) {
+    if (prop === 'providerData') {
+      return {};
+    } else {
+      return Reflect.get(...arguments);
+    }
+  },
+
+};
+
+/**
+ * Run all provisioning logic (e.g. Providers and Bidding Strategies)
+ */
+class Provisioner extends WMObject {
+  constructor({iterationGap=30, providers, biddingStrategies, datastore}) {
+    super({id: 'provisioner'});
+    this.providers = providers;
+    this.biddingStrategies = biddingStrategies;
+    this.datastore = datastore;
+    this.iterationGap = iterationGap;
+
+    this.iterate = new Iterate({
+      maxFailures: 30, // We really don't want it to crash
+      maxIterationTime: 300,
+      watchDog: 301,
+      waitTime: iterationGap,
+      handler: async () => {
+        await this.provision();
+      },
+    });
+    // TODO:
+    // make sure that iteration failures have a handler that's useful
```

Comment:
Is this just calling `monitor.reportError`?

Reply:
Possibly, but also possibly not, but that'll be implemented later.  More than likely, there will be some sort of error classification, then some will be sent to monitor.reportError.